added support for place holders in RSS savePath

ATGardner opened the pull request:
This is my 2nd attempt at creating this pull request. I apologize if I was supposed to somehow overwrite the [previous one](https://github.com/qbittorrent/qBittorrent/pull/5028), but I figured it will be easier to just create a new one.

I don't know how and where to add a UI explanation of the feature. I was thinking about maybe a tool tip letting the user know that he can use placeholders in "<" and ">", from the RSS entry's fields.

evsh commented on src/base/rss/rssfeed.cpp:
```diff
@@ -399,6 +399,32 @@ void Feed::downloadArticleTorrentIfMatching(const ArticlePtr &article)
+//          result = "c:\downloads\some attr value - some description\something\some title\not_found_attr"
+QString Feed::generateSavePath(const QString &savePathTpl, const ArticlePtr &article)
+{
+    QRegExp rx("<(.+)>");
```
Why such unusual syntax (<>)? Why don't you use something familiar? For instance:
```
 savePathTpl = /some/path/$ATTR{attr_name}/another/path/$ATTR{another_attr_name:or_default_value}
```
or
```
 savePathTpl = /some/path/${attr_name}/another/path/${another_attr_name:or_default_value}
```
or
```
 savePathTpl = /some/path/%{attr_name}/another/path/%{another_attr_name:or_default_value}
```

ATGardner replied:
All seem like good options. The exact placeholder syntax can be easily changed, of course.
I just picked something simple that works. I haven't thought about all the pros and cons of each and every option, and I wouldn't mind it being changed into anything that works well.

The addition of a default value is of course another good idea. That can be added as another feature.

ATGardner commented:
Is this PR even considered for a merge? I really enjoy this feature on my own machine, and I think other people would find it helpful as well.

glassez commented:
@ATGardner, are you still interested in this? Then you have to rebase it on top of current master. The RSS subsystem is redesigned now.

ATGardner commented:
@glassez - thanks. I'll get on it soon, and rebase on the new RSS design.

glassez commented:
@ATGardner, please, don't change Article class as you done in current version of this PR. If you need access fields by name implement it outside the class or use already implemented serialization feature (I mean `Article::toJsonObject()`).

ATGardner commented:
OK. I'll give it a shot.

ATGardner commented:
I got most of it to work, but using the toJsonObject only gives me the keys that are currently being saved in the Article instance, when in my example I want to use some other fields that the original RSS holds. That is the reason I initially changed the Article class to just hold a reference to the QVariantHash. This way I can extract any value from it that the user wishes to use in his savePath template.
Is there a simple way to get the original QVariantHash of the article without saving it in the instance?

glassez commented:
>That is the reason I initially changed the Article class to just hold a reference to the QVariantHash. This way I can extract any value from it that the user wishes to use in his savePath template.

It is inefficient to store QVariantHash, and convert every time someone gets some property of the Article object.
In this case, implement it outside the class.
>I got most of it to work, but using the toJsonObject only gives me the keys that are currently being saved in the Article instance, when in my example I want to use some other fields that the original RSS holds.

It has all keys, but some of them has new names (e.g. "torrent_url" -> "torrentURL").

ATGardner commented:
The problem isn't with the different name of torrentURL. It's with the missing properties. It only has id/date/title/author/description/torrentUrl/link/isRead. While my RSS feed contains many more custom fields I want to use for the template, like <tv:show_name>, or any other arbitrary field in the article.

If the problem if efficiency, I can just add a private member that keeps a reference to the QVariantHash, and only use it from a getValue function used by the AutoDownloadRule.
It will also have to be stored in the Article toJsonObject/fromJsonObject to have it read from the local files. Would converting the QVariantHash to QJsonObject be preferable, to make the storing and reading easier? 

Is that even a valid solution? I will make the local files basically much bigger, containing all the data that was originally in the feed item. Also - they will now contain some redundant data. I can change the Article's to only save the QVariantHash/QJsonObject of the original feed, and populate the other fields while reading from file (or even just lazily, when first called from the code, if it will make any difference)

glassez commented:
@ATGardner, I hear you! You need some additional article data. So I can suggest you use QObject Properties system. You can turn existing Article properties to QObject properties using Q_PROPERTY macro (of course, you shouldn't do it for "read" property since you really not need it, IIRC). Any additional data can be added as dynamic properties using `QObject::setProperty()`. Then you can access all Article properties (both static and dynamic) by its name using `QObject::property()`. Other components still can access it directly (as is now). What do you say?
>It will make the local files basically much bigger, containing all the data that was originally in the feed item.

I don't think we should worry too much about it. In any case it is not so much data. But I would still have done the uploading/storage of additional data to be optional (configurable via application settings).

ATGardner commented:
OK. I'll look into this QObject Properties idea. I will just create an article with dynamic properties matching whatever is in the original RSS item, and later access it when generating the final savePath.
Thanks for the direction.

ATGardner commented:
I admit I don't understand why adding those `Q_PROPERTY` macros would help, or why adding dynamic properties using `QObject::setProperty()` matters.
All I think I need to do is keep a reference to the original `QVariantHash` (probably converting it to QJsonObject - it seems easier to handle), in addition to populating the 8 standard Article properties, and add another `QString Article::getValue(QString name)` fucntion to it, so that the `generateSavePath` can extract all the needed data from it.
Why would adding a property be any better/easier?

glassez commented:
Because really this is what you need. And it already exists. I do not understand what the problem with that? At least it's more efficient than your method. 
See for yourself, most other components access the properties of the Article using its methods, but you propose to retrieve data from the variant hash each time (search + conversion) just because it suits you in one function, which may not be necessary at all.

glassez commented:
Of course, you can use your approach with `QString Article::getValue(QString name)` but you shouldn't use stored hash to access main properties.

ATGardner commented on src/base/rss/private/rss_parser.cpp:
```diff
@@ -280,36 +280,39 @@ void Parser::parseRssArticle(QXmlStreamReader &xml)
-            if (xml.name() == "title") {
-                article["title"] = xml.readElementText().trimmed();
+            const QString text(xml.readElementText().trimmed());
+            article[name] = text;
```
This is the main difference - the article QVariantHash will contain all of the feed item's fields

ATGardner commented on src/base/rss/private/rss_parser.cpp:
```diff
@@ -353,17 +356,21 @@ void Parser::parseAtomArticle(QXmlStreamReader &xml)
-            if (xml.name() == "title") {
-                article["title"] = xml.readElementText().trimmed();
+            const QString text(xml.readElementText().trimmed());
+            article[name] = text;
```
Same as line 290

ATGardner commented on src/base/rss/rss_article.cpp:
```diff
@@ -49,18 +51,18 @@ const QString Str_Read(QStringLiteral("read"));
+    , m_torrentURL(varHash.value(Str_Torrent_Url).toString())
+    , m_link(varHash.value(Str_News_Link).toString())
+    , m_isRead(varHash.value(Str_Read, false).toBool())
+    , m_varHash(varHash)
```
Save a reference to the hash

glassez commented on src/base/rss/rss_autodownloader.cpp:
```diff
@@ -248,25 +243,23 @@ void AutoDownloader::processJob(const QSharedPointer<ProcessingJob> &job)
-        params.addPaused = rule.addPaused();
-        BitTorrent::Session::instance()->addTorrent(job->torrentURL, params);
+        if (auto feed = Session::instance()->feedByURL(job->feedURL)) {
+            if (auto article = feed->articleByGUID(job->articleGUID)) {
```
The auto downloading jobs are performed deferred, so article as well as its feed might not exist at this point, but we still have to handle the job.
Please, try first to understand how it works.

glassez replied:
Maybe you don't need `Article::value()` but just store JSON object in downloading job instead of currently used article fields?

glassez commented on src/base/rss/rss_autodownloader.cpp:
```diff
@@ -248,25 +243,23 @@ void AutoDownloader::processJob(const QSharedPointer<ProcessingJob> &job)
+        if (auto feed = Session::instance()->feedByURL(job->feedURL)) {
+            if (auto article = feed->articleByGUID(job->articleGUID)) {
+                BitTorrent::AddTorrentParams params;
+                params.savePath = AutoDownloader::generateSavePath(rule, article);
```
AutoDownloader:: is redundant here.

ATGardner commented on src/base/rss/rss_article.cpp:
```diff
@@ -104,6 +106,10 @@ bool Article::isRead() const
 }
 
+QString Article::value(const QString key) const {
+    return m_varHash.value(key, QString('_')).toString();
```
Just an empty string would cause the download to fail (converting '/folder/<placeholder>/whatever' to '/folder//whatever', for example). I wasn't sure how it would handle a blank space (' '). But I figured an underscore is easy enough to see. It can also just be the value of the "key" parameter.

LordNyriox replied:
@glassez, @ATGardner:  Also, some filesystems do not handle spaces very well, often requiring additional quotes, slashes/backslashes, and the like.

glassez commented on src/base/rss/private/rss_parser.cpp:
```diff
@@ -280,36 +280,30 @@ void Parser::parseRssArticle(QXmlStreamReader &xml)
 
     while (!xml.atEnd()) {
         xml.readNext();
+        const QString name(xml.name().toString());
```
xml.name() isn't string?

glassez commented on src/base/rss/rss_article.cpp:
```diff
@@ -114,16 +115,9 @@ void Article::markAsRead()
-        {Str_Link, m_link},
-        {Str_IsRead, m_isRead}
-    };
+    QJsonObject jsonObj(QJsonObject::fromVariantHash(m_varHash));
```
Don't try to cheat it! QJsonObject::fromVariantHash doesn't support all QVariants (at least it doesn't support QDateTime).

glassez commented:
>I understand you have problems because RSS subsystem was changed recently and many things are still evolving. I can offer you the following: I will make the necessary changes in the code of the RSS::Parser and RSS::Article, and you then change your PR to contain just the main feature - generating save path.
What do you say?

@ATGardner, OK. I see you have a little time to do so. I'll change something myself in the near future, so wait here.

ATGardner commented:
I am not sure how I can separate the changes to the parser code, and what you want to do with the constants. Would it be easier for you to just have access to my fork branch, and do your thing there? Or do you prefer doing it on the master, and I'll rebase on it, and then just add the generateSavePath?

glassez commented:
>Or do you prefer doing it on the master, and I'll rebase on it, and then just add the generateSavePath?

Yes. I prefer this way.
I have already started working on it. I will notify you when the updated code is in the master.

glassez commented:
@ATGardner, please review #6864.

ATGardner commented:
It looks good. Thanks for your help on the matter.
Should I wait until it is merged to the master, and then pull to my fork, and rebase on it?

glassez commented:
#6864 merged so you can rebase.

ATGardner commented:
I rebased, and I think I implemented the generateSavePath function again. But ever since I upgraded to VS2017, I can't compile qBittorrent at all.
I tried following the instructions in the wiki for MSVC 2017, but my qt creator keeps complaining about not being able to find different members on various QT classes. stuff like
...\qBittorrent\src\gui\mainwindow.cpp:300: error: C2039: 'actionShowStatusbar': is not a member of 'Ui::MainWindow'

I couldn't find any help on how to compile properly, so I'm a bit stuck now. I can try to commit and push my changes, but I am not sure that they are 100% working currently.

glassez commented:
@ATGardner, apparently, are you lost in these rebasings. It would be easier for you to hard reset your branch to qBittorrent master (you drop all your changes doing it so save needed code in some place before) and create commit with your feature again.

ATGardner commented:
No, the rebase went fine. I think I managed to re-implement quite easily. It just won't compile on my machine anymore, not related to the changes I've made. So I can't test my code to see if it works.

It's really a painful process to compile all the zlib, openssl, boost, libtorrent and qt5, just so it can be statically linked to the project. But I think I followed all the instructions, and yet QT Creator just won't compile, with about 260 errors about different methods in QT, in areas I have not touched at all.

Up until several weeks ago, I used VS2015, and I somehow managed to get the whole fragile project to compile a long time ago, and haven't touched it since (just rebuild and relaunch when I changed the RSS code files). But since moving to VS2017 - no luck.

Is there anybody on the qbittorrent team which actually uses the VS2017 method, and might give me a few pointers as to what I'm doing wrong?

glassez commented on src/base/rss/rss_autodownloader.h:
```diff
@@ -99,6 +99,8 @@ namespace RSS
         void store();
         void storeDeferred();
 
+        static QString generateSavePath(const QString savePath, const QVariantHash articleData);
```
generateSavePath shouldn't be static member of AutoDownloader. Change it to free function in anonymous namespace inside "rss_autodownloader.cpp".

glassez commented on src/base/rss/rss_autodownloader.cpp:
```diff
@@ -52,6 +52,36 @@
 #include "rss_folder.h"
 #include "rss_session.h"
 
+namespace {
```
Fix coding style.

evsh commented on src/base/rss/rss_autodownloader.cpp:
```diff
@@ -52,6 +52,36 @@
+    //                      <description>some description</description>
+    //                      <attr_name>some attr value</attr_name>
+    //                    </item>
+    //          result = "c:\downloads\some attr value - some description\something\some title\not_found_attr"
```
Reading the code below can't understand how does the string  "not_found_attr" appear.

ATGardner replied:
I wrote the comment when I used
QString value = articleData.value(key, key).toString();
But I later changed it. I guess I should have kept it in, or updated the comment to reflect it.

evsh commented on src/base/rss/rss_autodownloader.cpp:
```diff
@@ -52,6 +52,36 @@
+    //          result = "c:\downloads\some attr value - some description\something\some title\not_found_attr"
+    QString generateSavePath(const QString savePath, const QVariantHash articleData)
+    {
+        QRegExp rx("<(.+)>");
```
Don't like this syntax because it is hardly extensible. For example, how would you add default values? One has to reserve a set of characters for the special purpose like denoting a default value or value prefixes/affixes. If we release this free form into the wild life, users create directory structures with all possible characters in names and we will have problems in the future extending the rule syntax.

ATGardner replied:
I just picked the simplest syntax that would get my feature working for my use cases. Of course it can be revised and expanded, to include more complicated ones.
I am not an expert in regexp, nor am I a c++ programmer. Adding to that - I can't seem to compile the project currently on my computer, using MSVC2017 (is there anybody that can help me with that?). But if you have any specific ideas, I can give it a shot.

evsh commented on src/base/rss/rss_autodownloader.cpp:
```diff
@@ -52,6 +52,36 @@
+        rx.setMinimal(true);
+        QString result = savePath;
+
+        int pos = 0;
```
Can't you declare 'pos' inside of the loop scope? For example:
```c++
for (int pos = 0; pos != -1; pos = rx.indexIn(result, pos))) {
```
